pkg/controllers/provisioning/provisioner.go

```diff
@@ -47,6 +47,8 @@ var (
 )
 
 func NewProvisioner(ctx context.Context, provisioner *v1alpha5.Provisioner, kubeClient client.Client, coreV1Client corev1.CoreV1Interface, cloudProvider cloudprovider.CloudProvider) *Provisioner {
+	provisioner = provisioner.DeepCopy()
+	provisioner.SetDefaults(ctx)
```

Comment:
I'm not sure how comfortable I am with runtime defaults. What's the advantage of this vs just defaulting this where it's used?

Reply:
I think my main concern is DRYing defaulting logic. If we don't need to support defaulting metadata option fields during mutating webhook then the defaulting for metadata options fields can probably go elsewhere.

Of course, if we don't default the new fields with a mutating webhook then we also have to loosen some of the validation that will run during the validating webhook (which currently requires that the `metadataOptions` field is non-nil). Looser validation might be acceptable though.

Reply:
My thinking was that new fields are all optional, but they are defaulted and validated at the API server (via webhooks). That gives the advantage of rejecting bad objects synchronous with the user. It also more clearly communicates to the user what the values will be when defaulted (as they become visible in the API upon creation).

The main disadvantage is when updating the controller to a version that includes this new feature. Any objects previously created in the API were not defaulted yet. So the controller would essentially be reconciling invalid objects. By defaulting again here we could fix all previously created objects in the API in-memory as part of reconciliation.

I suppose that by defaulting only where the metadataOption fields are consumed we can side-step the issue of invalid objects previously stored in the API. I agree that might be a little cleaner and we'd just have to lean more on documentation to accurately communicate default values to the user. Let me give that approach a go.